### PR TITLE
[MLIR][OpenMP][Offload] Lower target update op to DeviceRT

### PR DESCRIPTION
**Pull request by ergawy**
Adds support for lowring `UpdateDataOp` to the DeviceRT. This reuses the
existing utils used by other device directive.

### CODE REVIEW

File: mlir/test/Target/LLVMIR/omptarget-llvm.mlir [hunk 1]
```diff
@@ -441,3 +441,41 @@ llvm.func @_QPopenmp_target_use_dev_both() {
...
+  %0 = llvm.mlir.constant(1 : i64) : i64
+  %1 = llvm.alloca %0 x i32 {bindc_name = "i", in_type = i32, operand_segment_sizes = array<i32: 0, 0>, uniq_name = "_QFopenmp_target_dataEi"} : (i64) -> !llvm.ptr
+  %2 = omp.map_info var_ptr(%1 : !llvm.ptr, i32)   map_clauses(tofrom) capture(ByRef) -> !llvm.ptr {name = ""}
+  omp.target_data map_entries(%2 : !llvm.ptr) {
+    %3 = llvm.mlir.constant(99 : i32) : i32
+    llvm.store %3, %1 : i32, !llvm.ptr
+    omp.terminator
+  }
+
+  omp.target_update_data motion_entries(%2 : !llvm.ptr)
```
Comment (TIFitis) [inline mlir/test/Target/LLVMIR/omptarget-llvm.mlir:455]: You can't reuse `%2` here, `tofrom` is not a valid map_type for target update. This check should also be added to the verifier.

Comment (ergawy) [reply mlir/test/Target/LLVMIR/omptarget-llvm.mlir:455]: That's true. I added the check to the verifier in the parent PR.

File: mlir/test/Target/LLVMIR/omptarget-llvm.mlir [hunk 2]
```diff
@@ -441,3 +441,41 @@ llvm.func @_QPopenmp_target_use_dev_both() {
 // CHECK:         ret void
 
 // -----
+
+llvm.func @_QPopenmp_target_data_update() {
```
Comment (DominikAdamski) [inline mlir/test/Target/LLVMIR/omptarget-llvm.mlir:445]: Could you add more tests to cover different data types (like Fortran pointers, allocatable arrays etc.)

**Review by TIFitis — APPROVED** [review]
LGTM, Thanks for the patch!
Please make sure build bots are clean before merging 👍🏽

File: mlir/test/Target/LLVMIR/omptarget-llvm.mlir (hunk 2 above)
Comment (ergawy) [reply mlir/test/Target/LLVMIR/omptarget-llvm.mlir:445]: As discussed offline, the FIR dialect is not registered with `mlir-translate`. So we cannot use `fir.ref` in `omptarget-llvm.mlir`.

However, I will double check with @agozillon whether we have existing tests for lowering `map_info` op's arguments from `fir` to `llvm` types or not. If not, I will open another PR with more testing.

**Pull request merged**